RecoPixelVertexing/PixelTriplets/plugins/gpuPixelDoublets.h

```diff
@@ -20,157 +20,168 @@ namespace gpuPixelDoublets {
+        for (auto p = hist[outer].begin(kk); p < hist[outer].end(kk); ++p) {
+          auto oi=*p;
+          assert(oi>=offsets[outer]);
+          assert(oi<offsets[outer+1]);
+
+          if (std::min(std::abs(int16_t(iphi[oi]-mep)), std::abs(int16_t(mep-iphi[oi]))) > iphicut)
+            continue;
+          if (z0cutoff(oi) || ptcut(oi)) continue;
+          auto ind = atomicInc(nCells, MaxNumOfDoublets);
+          // int layerPairId, int doubletId, int innerHitId, int outerHitId)
+          cells[ind].init(hh, pairLayerId, ind, i, oi);
+          isOuterHitOfCell[oi].push_back(ind);
+          if (isOuterHitOfCell[oi].full()) ++tooMany;
+          ++tot;
+        }
       }
-    }
-    if (tooMany>0) printf("OuterHitOfCell full for %d in layer %d/%d, %d:%d   %d,%d\n", i, inner,outer, kl,kh,nmin,tot);
-
-    if (hist[outer].nspills>0)
-      printf("spill bin to be checked in %d %d\n",outer,hist[outer].nspills);
+      if (tooMany > 0)
+        printf("OuterHitOfCell full for %d in layer %d/%d, %d:%d   %d,%d\n", i, inner, outer, kl, kh, nmin, tot);
 
-    // if (0==hist[outer].nspills) assert(tot>=nmin);
-    // look in spill bin as well....
+      if (hist[outer].nspills > 0)
+        printf("spill bin to be checked in %d %d\n", outer, hist[outer].nspills);
 
+      // if (0==hist[outer].nspills) assert(tot>=nmin);
+      // look in spill bin as well....
 
-  }  // loop in block...
+    }  // loop in block...
   }
 
   __global__
-  void getDoubletsFromHisto(GPUCACell * cells, uint32_t * nCells, siPixelRecHitsHeterogeneousProduct::HitsOnGPU const *  __restrict__ hhp,
-                            GPU::VecArray< unsigned int, 256> *isOuterHitOfCell) {
-
-    uint8_t const layerPairs[2*13] = {0,1 ,1,2 ,2,3
-                                     // ,0,4 ,1,4 ,2,4 ,4,5 ,5,6
-                                     ,0,7 ,1,7 ,2,7 ,7,8 ,8,9
-                                     ,0,4 ,1,4 ,2,4 ,4,5 ,5,6
-                                     };
-
-    const int16_t phi0p05 = phi2short(0.05);
-    const int16_t phi0p06 = phi2short(0.06);
-    const int16_t phi0p07 = phi2short(0.07);
+  void getDoubletsFromHisto(GPUCACell * cells,
+                            uint32_t * nCells,
+                            siPixelRecHitsHeterogeneousProduct::HitsOnGPU const *  __restrict__ hhp,
+                            GPU::VecArray<unsigned int, 256> * isOuterHitOfCell)
+  {
+    constexpr int nPairs = 13;
+    constexpr const uint8_t layerPairs[2*nPairs] = {
+      0, 1,  1, 2,  2, 3,
+   // 0, 4,  1, 4,  2, 4,  4, 5,  5, 6,
+      0, 7,  1, 7,  2, 7,  7, 8,  8, 9,
+      0, 4,  1, 4,  2, 4,  4, 5,  5, 6
+    };
 
-    int16_t const phicuts[13] { phi0p05, phi0p05, phi0p06
-                               ,phi0p07, phi0p06, phi0p06, phi0p05, phi0p05
-                               ,phi0p07, phi0p06, phi0p06, phi0p05, phi0p05
-                              };
+    constexpr int16_t phi0p05 = 522;    // round(521.52189...) = phi2short(0.05);
```

Comment:
in principle phi2short should be computed at compile time

Reply:
I tried, and the compiler said it is not able to, because `std::round()` and similar are not `constexpr` :-(

Reply:
ah yes, same for vectorization on x86. we have a constexpr (fast) version (in C) of round somewhere